### ISSUE TITLE
WinUI 3: LinearGradientBrush and RadialGradientBrush don't both inherit from GradientBrush

### ISSUE THREAD

**Issue by michael-hawker** [microsoft/microsoft-ui-xaml#3883]
**Describe the bug**
In WPF both [`LinearGradientBrush`](https://docs.microsoft.com/en-us/dotnet/api/system.windows.media.lineargradientbrush?view=net-5.0) and [`RadialGradientBrush`](https://docs.microsoft.com/en-us/dotnet/api/system.windows.media.radialgradientbrush?view=net-5.0) inherit from the [`GradientBrush`](https://docs.microsoft.com/en-us/dotnet/api/system.windows.media.gradientbrush?view=net-5.0) base class. In WinUI this is not the case. I can understand in WinUI 2 that this was an added feature, but with WinUI 3 and access to modify the underlying structures, it seems like better alignment should be achievable.

Having them inherit from the same base class makes it easier to maintain consistency and more easily switch brushes between linear and radial variants by adding the additional parent properties. It also means that library owners can create extensions or helpers that target the base `GradientBrush` class that will work with both classes (for instance Animation helpers).

**Expected behavior**
Both Gradient Brushes should share the same base class.

**Version Info**

NuGet package version: Microsoft.WinUI 3.0.0-preview3.201113.0

**Comment by michael-hawker** [comment]
In the meantime, it'd be nice if the `GradientStops` property was a `GradientStopCollection` explicitly vs. the `IObservableVector<GradientStops>`?

**Comment by StephenLPeters** [comment]
@jesbis Do you know why RadialGradientBrush wasn't a GradientBrush in the first place? @MikeHillberg fyi?

**Comment by jesbis** [comment]
It couldn't be a `GradientBrush` in WinUI 2.x since brushes other than `XamlCompositionBrushBase` aren't meaningfully extensible - there's no way to render custom brushes.

With WinUI 3 it would theoretically be possible to update the rendering pipeline and redo the brush class hierarchy. 

Although maybe worth considering: what is the desired extensibility model for brushes overall? Would **all** brushes ideally be a `XamlCompositionBrushBase` instead?

**Comment by michael-hawker** [comment]
In some ways this is to raise awareness about a general issue about making sure the API surface is being unified as WinUI 2 and 3 come together. But this happens to just be a concrete example to show the point.

Yeah @jesbis if the `XamlCompositionBrushBase` is the best new base class for these brushes that's fine. `GradientBrush` could be built on top of that instead going in the other direction so that they have the same shared properties. 🙂

**Comment by MikeHillberg** [comment]
This is a result of
* we're trying to keep WinUI3 APIs compatible with WinUi2 (caveat the namespace change)
* we expected the polymorphism to be based primarily on Brush rather than GradientBrush (and they share the Brush base)

**Comment by michael-hawker** [comment]
@MikeHillberg but this means that unlike WPF there's no common pattern between the two brushes. Nor have their properties at least been implemented similarly (i.e. using `GradientStopCollection`). They both have a collection of `GradientStops` but implement them in entirely different ways. 

This means there's no way to extend or write a helper that easily can help generate or manipulate gradient stops for both types of brushes.

From the compatibility perspective, the major version is the opportunity and time to make adjustments for improvements. Especially for some of these things in most cases (especially from XAML usage) there won't be a compatibility break.

From the polymorphism perspective, in WinUI 2 world the underlying type difference made sense as there wasn't another way to implement it outside the OS.

I just worry that if some things aren't adjusted/corrected now, it'll be years before a 'WinUI 4' where these things can be fixed.

**Comment by MikeHillberg** [comment]
Yeah, I get it. It's a tradeoff and between easing migration and taking this as an opportunity to make some incompatible improvements. The decision we made for WinUI 3.0 is to be compat with WinUI2 wherever possible, because we thought it was important, and because there are some unavoidable incompatibilities that will already bring us below 100%.

It's too late to revisit that decision now for 3.0. Looking ahead, timing of a 4.0 will be (is) another fun topic.

**Comment by michael-hawker** [comment]
Bump, related to #7967 as well.

**Comment by kmahone** [comment]
This isn't something that we are going to be able to change. LinearGradientBrush and RadialGradientBrush do not share any implementation. LinearGradientBrush is implemented natively in Microsoft.UI.Xaml.dll, RadialGradientBrush is implemented in Microsoft.UI.Xaml.Controls.dll and is based on XamlCompositionBrush. Creating a new GradientBrush base class would require reimplementing one or both of these brushes from scratch. There would be some nice benefits to do this, but not enough to justify the cost.
